### PR TITLE
Add archive

### PR DESCRIPTION
**Pull request by yvesdaxmaz**
## Pullrequest

Allow to archive mail by moving it in an archive folder, If the archive folder doesn't exists it's create first time a message is been archived.

### Issues

- [X] Archive message #228

### Checklist

- [X] None

### How2Test

- [X] Open any message and click on the Archive button in message header
- [X ] Verify if Archive folder have been created and the archived message is in that folder

### Todo

- [X] Create submailbox with the name of the mailbox where the message was before he been archived and place the message and that mailbox and not in the Archive message.
When I try to create a submailbox I can't get access to the real name of the mailbox where the mail was before he been archived instead of that I am getting some random sequence of character which is completely differente from the name which is in the list of mailbox

### CODE REVIEW

**Comment by jasonmunro** [comment]
@yvesdaxmaz I will review this as soon as I can, probably this weekend. The folder name is "encoded" to make it URL friendly with bin2hex. You can "decode" it to get the actual folder name with hex2bin().

**Comment by yvesdaxmaz** [comment]
@jasonmunro thanks, I will perform change  today and you can review all at once

**Comment by yvesdaxmaz** [comment]
I have extend archive feature this by adding in folder settings possibility to set the archive folder.

**Comment by jasonmunro** [comment]
Sorry for the late reply (again). Will review and merge as soon as I can. Thanks for doing this!

File: modules/imap/handler_modules.php [hunk 1]
```diff
@@ -685,6 +685,82 @@ public function process() {
...
+                    }
+                }
+            }
+
+            if(!$archive_folder) {
+                $archive_folder = "Archive";
+            }
+
+            if (imap_authed($imap)) {
+                $new_folder = $archive_folder . '/' . hex2bin($form['folder']);
```
[the remark this answers is not in the document]
Comment (jasonmunro) [reply modules/imap/handler_modules.php]: ok, $imap->delim did NOT fix the issue :) it must be set to false which caused the 2 folders to be created in my case. I will dig more into it.

[the remark this answers is not in the document]
Comment (jasonmunro) [reply modules/imap/handler_modules.php]: pretty sure we need to call $imap->get_namespaces() to get the delimiter, and then this will be good to go. I will let follow up as soon as I can.

[the remark this answers is not in the document]
Comment (yvesdaxmaz) [reply modules/imap/handler_modules.php]: I made update to use delimiter used by the imap server

[the remark this answers is not in the document]
Comment (jasonmunro) [reply modules/imap/handler_modules.php]: Looks good! I am going to merge this as is then make a few changes, specifically:
- we have a function to properly format a folder for create_mailbox(), however it lives in the imap_folders module so I'm going to move it to the imap module and use it
- when using caching new folders are not showing up without logging out and back in. This is unrelated to your changes and I just fixed it in master for the imap_folders module set, so I will apply that same fix here.
- the logic that creates the folders then moves the message is a bit hard to follow, I'm going to do some refactoring there to make it easier to understand but it won't change the functionality.
- Have you tried archiving a message in a subfolder, that is in another subfolder? like FOO/BAR/BAZ? I'm curious what will happen there. I think it might not work but have not tested it yet.

Overall great work and thanks for doing this!

**Comment by marclaporte** [comment]
I am using this every day. It works well and is very useful. Thanks @yvesdaxmaz and @jasonmunro !

**Comment by jasonmunro** [comment]
@marclaporte Awesome! @yvesdaxmaz deserves thanks for the great work!

**Comment by marclaporte** [comment]
For the record, when I use, it doesn't archive messages to the "Original folder" as I requested above, so I just opened a spin off task: https://github.com/jasonmunro/cypht/issues/571
